romancal/resample/resample_step.py

```diff
@@ -68,43 +68,49 @@ class ResampleStep(RomanStep):
 
     def process(self, input):
         if isinstance(input, datamodels.DataModel):
-            input_models = ModelContainer([input])
+            input_models = ModelLibrary([input])
             # set output filename from meta.filename found in the first datamodel
-            output = input_models[0].meta.filename
+            output = input.meta.filename
             self.blendheaders = False
         elif isinstance(input, str):
             # either a single asdf filename or an association filename
             try:
                 # association filename
-                input_models = ModelContainer(input)
+                input_models = ModelLibrary(input)
             except Exception:
                 # single ASDF filename
-                input_models = ModelContainer([input])
-            if hasattr(input_models, "asn_table") and len(input_models.asn_table):
-                # set output filename from ASN table
-                output = input_models.asn_table["products"][0]["name"]
-            elif hasattr(input_models[0], "meta"):
-                # set output filename from meta.filename found in the first datamodel
-                output = input_models[0].meta.filename
-        elif isinstance(input, ModelContainer):
+                input_models = ModelLibrary([input])
+            # FIXME: I think this can be refactored and maybe could be common code
+            # for several steps
+            output = input_models.asn["products"][0]["name"]
+            # if hasattr(input_models, "asn_table") and len(input_models.asn_table):
+            #     # set output filename from ASN table
+            #     output = input_models.asn_table["products"][0]["name"]
+            # elif hasattr(input_models[0], "meta"):
+            #     # set output filename from meta.filename found in the first datamodel
+            #     output = input_models[0].meta.filename
+        elif isinstance(input, ModelLibrary):
             input_models = input
             # set output filename using the common prefix of all datamodels
-            output = (
-                f"{os.path.commonprefix([x.meta.filename for x in input_models])}.asdf"
-            )
-            if len(output) == 0:
+            # TODO can this be set from the members?
+            output = f"{os.path.commonprefix([x['expname'] for x in input_models.asn['products'][0]['members']])}.asdf"
+            if len(output) == 0:  # FIXME won't this always at least be ".asdf"?
                 # set default filename if no common prefix can be determined
                 output = "resample_output.asdf"
         else:
             raise TypeError(
-                "Input must be an ASN filename, a ModelContainer, "
+                "Input must be an ASN filename, a ModelLibrary, "
                 "a single ASDF filename, or a single Roman DataModel."
             )
```

Comment:
This is another example that would benefit from letting `ModelLibrary` handle the input type.

Reply:
Looking at this again I have questions about the refactoring here and https://github.com/spacetelescope/romancal/pull/1241#discussion_r1629493747

The tldr; is that I don't think further refactoring is needed for `ModelLibrary` and the cleanup of the input handling could be in a separate PR.

The flux step input is relatively straightforward. However resample is complicated by a few things:
- `blendeheaders` is only set if the input is `DataModel` (and not if the input a filename for a single datamodel)
- the `output` for a `ModelContainer` is set based on the common filename whereas the `output` for a `ModelContainer` read from an association file is based on the product name

Any refactoring of the input handling will either need to replicate these differences (and won't result in significant or possibly any clean up) or resolve the differences above (which would require more discussion and possibly test changes). Given these issues I think these changes warrent a separate PR (and aren't required by the `ModelLibrary`).

Reply:
FYI, I think blendheaders is just a Webb holdover and we have done basically everything we need to for the L3 metadata.  i.e., we could just get rid of blendheaders.

The single datamodel resample case I don't think is one we need to worry a lot about; if we want to change the behavior there to support a simpler code flow, we could do that.  But I agree we can take that up in a separate PR.

Reply:
I opened an issue to track the input handling logic: https://github.com/spacetelescope/romancal/issues/1309